File: bench/ts/src/bench-jed.ts
// bench-jed benchmarks the TS jed core (spec/design/benchmarks.md §6/§7). The core is
// imported relatively — Node's native type-stripping runs it with no build step, and
// this package's dependencies never touch impl/ts.

import { mkdtempSync, rmSync } from "node:fs";
import { join } from "node:path";

import {
  type Engine as JedDb,
  type PreparedStatement,
  type Session,
  create,
  executeParams,
  executePrepared,
  open,
  openDatabase,
  prepare,
  query,
  queryPrepared,
} from "../../../impl/ts/src/tooling.ts";
import { intValue, textValue, type Value } from "../../../impl/ts/src/value.ts";

import {
  type Arg,
  Checksum,
  type ConcurrentOutcome,
  type Engine,
  mainWith,
  readSidecar,
} from "./lib.ts";

class JedEngine implements Engine {
  private readonly db: JedDb;
  private stmt: PreparedStatement | null = null;
  private readonly dataDir: string;
  private readonly dataset: string;
  private readonly scratch: string | null;

  constructor(db: JedDb, dataDir: string, dataset: string, scratch: string | null) {
    this.db = db;
    this.dataDir = dataDir;
    this.dataset = dataset;
    this.scratch = scratch;
  }

  async exec(sql: string): Promise<void> {
    executeParams(this.db, sql, []);
  }

  async prepare(sql: string): Promise<void> {
    this.stmt = prepare(this.db, sql);
  }

  async queryPrepared(args: Arg[], sum: Checksum | null): Promise<number> {
    const rows = queryPrepared(this.db, this.stmt!, bindArgs(args));
    let n = 0;
    for (const row of rows) {
      n++;
      if (sum === null) continue;
      for (const v of row) {
        switch (v.kind) {
          case "null":
            sum.null();
            break;
          case "int":
            sum.int(v.int);
            break;
          case "text":
            sum.text(v.text);
            break;
          default:
            throw new Error(`unexpected result kind ${v.kind}`);
        }
      }
      sum.endRow();
    }
    return n;
  }

  async execPrepared(args: Arg[]): Promise<void> {
    executePrepared(this.db, this.stmt!, bindArgs(args));
  }

  async queryInt(sql: string): Promise<bigint> {
    for (const row of query(this.db, sql, [])) {
      const v = row[0];
      if (v.kind === "int") return v.int;
    }
    throw new Error(`expected one integer from ${sql}`);
  }

  async storedFingerprint(): Promise<string> {
    return readSidecar(this.dataDir, this.dataset, "jed");
  }

  async close(): Promise<void> {
    if (this.scratch !== null) {
      rmSync(this.scratch, { recursive: true, force: true });
    }
  }

  // concurrentRead opens ONE Database over the dataset file and mints a reader Session per
  // block (the slice-7 convergence, session.md §2.4/§10) — every Session shares the one
  // Database's committed snapshot + buffer pool. The single-threaded TS core runs the
  // blocks SEQUENTIALLY (no parallel speedup, documented), but the per-block partition +
  // fold makes the answer checksum identical to the threaded Go/Rust cores (benchmarks.md §8.1).
  async concurrentRead(
    sql: string,
    warm: Arg[][][],
    meas: Arg[][][],
    expectRows: number,
  ): Promise<ConcurrentOutcome> {
    const db = openDatabase(join(this.dataDir, `${this.dataset}.jed`));
    const readers = meas.length;
    const sessions: Session[] = Array.from({ length: readers }, () => db.readSession());
    try {
      // Pass 1 — warmup, untimed: populate the shared buffer pool.
      for (let r = 0; r < readers; r++) {
        for (const args of warm[r]) runReaderQuery(sessions[r], sql, args, null);
      }
      // Pass 2 — measured (wall-clock).
      const blockHexes: string[] = [];
      const elapsed: bigint[] = [];
      let rowsTotal = 0;
      const start = process.hrtime.bigint();
      for (let r = 0; r < readers; r++) {
        const sum = new Checksum();
        for (const args of meas[r]) {
          const t0 = process.hrtime.bigint();
          const n = runReaderQuery(sessions[r], sql, args, sum);
          elapsed.push(process.hrtime.bigint() - t0);
          rowsTotal += n;
          if (expectRows > 0 && n !== expectRows) {
            throw new Error(`expected ${expectRows} rows per iteration, got ${n}`);
          }
        }
        blockHexes.push(sum.hex());
      }
      const wallNs = process.hrtime.bigint() - start;
      return { blockHexes, elapsed, rowsTotal, wallNs };
    } finally {
      for (const s of sessions) s.close();
      db.close();
    }
  }
}

function bindArgs(args: Arg[]): Value[] {
  return args.map((a) => (typeof a === "bigint" ? intValue(a) : textValue(a)));
}

// runReaderQuery runs one query through a reader Session, re-parsing the SQL each call — deliberate
// (benchmarks.md §8.1): a constant per-query parse cost is included, uniform across the jed cores —
// folding rows into sum.
function runReaderQuery(sess: Session, sql: string, args: Arg[], sum: Checksum | null): number {
  const rows = sess.query(sql, bindArgs(args));
  let n = 0;
  for (const row of rows) {
    n++;
    if (sum === null) continue;
    for (const v of row) {
      switch (v.kind) {
        case "null":
          sum.null();
          break;
        case "int":
          sum.int(v.int);
          break;
        case "text":
          sum.text(v.text);
          break;
        default:
          throw new Error(`unexpected result kind ${v.kind}`);
      }
    }
    sum.endRow();
  }
  return n;
}

await mainWith({
  engine: "jed",
  lang: "ts",
  variant: "core",
  async open(dataDir: string, dataset: string): Promise<Engine> {
    if (dataset === "scratch") {
      const dir = mkdtempSync(join(dataDir, "scratch-"));
      return new JedEngine(create(join(dir, "scratch.jed")), dataDir, dataset, dir);
    }
    return new JedEngine(open(join(dataDir, `${dataset}.jed`)), dataDir, dataset, null);
  },
});

File: bench/rust/src/bin/bench_jed.rs
//! bench-jed benchmarks the Rust jed core (spec/design/benchmarks.md §6/§7).

use std::time::Instant;

use jed::{Database, DatabaseOptions, PreparedStatement, Session, SessionOptions, SharedCore, Value};
use jed_bench::{
    Arg, BoxResult, Checksum, ConcurrentOutcome, Config, Engine, main_with, read_sidecar,
};

fn main() {
    main_with(Config {
        engine: "jed",
        lang: "rust",
        variant: "core",
        open,
    });
}

struct JedEngine {
    // The persistent connection the bench drives (BEGIN/COMMIT/ROLLBACK span calls). It owns an
    // `Arc<Shared>`, so it keeps the storage alive after the local `Database` handle is dropped.
    sess: Session,
    stmt: Option<PreparedStatement>,
    data_dir: String,
    dataset: String,
    scratch: Option<String>, // temp dir holding the scratch file, removed on Drop
}

fn open(data_dir: &str, dataset: &str) -> BoxResult<Box<dyn Engine>> {
    if dataset == "scratch" {
        let dir = format!("{data_dir}/scratch-rust-{}", std::process::id());
        std::fs::create_dir_all(&dir)?;
        let db = Database::create(format!("{dir}/scratch.jed"), DatabaseOptions::default())
            .map_err(|e| e.to_string())?;
        let sess = db.session(SessionOptions::default());
        return Ok(Box::new(JedEngine {
            sess,
            stmt: None,
            data_dir: data_dir.to_string(),
            dataset: dataset.to_string(),
            scratch: Some(dir),
        }));
    }
    let db = Database::open(format!("{data_dir}/{dataset}.jed")).map_err(|e| e.to_string())?;
    let sess = db.session(SessionOptions::default());
    Ok(Box::new(JedEngine {
        sess,
        stmt: None,
        data_dir: data_dir.to_string(),
        dataset: dataset.to_string(),
        scratch: None,
    }))
}

impl Drop for JedEngine {
    fn drop(&mut self) {
        if let Some(dir) = &self.scratch {
            let _ = std::fs::remove_dir_all(dir);
        }
    }
}

fn bind_args(args: &[Arg]) -> Vec<Value> {
    args.iter()
        .map(|a| match a {
            Arg::Int(n) => Value::Int(*n),
            Arg::Text(s) => Value::Text(s.clone()),
        })
        .collect()
}

impl Engine for JedEngine {
    fn exec(&mut self, sql: &str) -> BoxResult<()> {
        self.sess.execute(sql, &[]).map_err(|e| e.to_string())?;
        Ok(())
    }

    fn prepare(&mut self, sql: &str) -> BoxResult<()> {
        self.stmt = Some(self.sess.prepare(sql).map_err(|e| e.to_string())?);
        Ok(())
    }

    fn query_prepared(&mut self, args: &[Arg], sum: Option<&mut Checksum>) -> BoxResult<usize> {
        let params = bind_args(args);
        let stmt = self.stmt.as_ref().expect("prepare first");
        let rows = self
            .sess
            .query_prepared(stmt, &params)
            .map_err(|e| e.to_string())?;
        let mut n = 0;
        match sum {
            None => {
                for _ in rows {
                    n += 1;
                }
            }
            Some(sum) => {
                for row in rows {
                    n += 1;
                    for v in row {
                        match v {
                            Value::Null => sum.null(),
                            Value::Int(x) => sum.int(x),
                            Value::Text(s) => sum.text(&s),
                            other => {
                                return Err(format!("unexpected result value {other:?}").into());
                            }
                        }
                    }
                    sum.end_row();
                }
            }
        }
        Ok(n)
    }

    fn exec_prepared(&mut self, args: &[Arg]) -> BoxResult<()> {
        let params = bind_args(args);
        let stmt = self.stmt.as_ref().expect("prepare first");
        self.sess
            .execute_prepared(stmt, &params)
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    fn query_int(&mut self, sql: &str) -> BoxResult<i64> {
        let mut rows = self.sess.query(sql, &[]).map_err(|e| e.to_string())?;
        match rows.next().and_then(|r| r.into_iter().next()) {
            Some(Value::Int(n)) => Ok(n),
            other => Err(format!("expected one integer from {sql:?}, got {other:?}").into()),
        }
    }

    fn stored_fingerprint(&mut self) -> BoxResult<String> {
        Ok(read_sidecar(&self.data_dir, &self.dataset, "jed"))
    }

    // concurrent_read opens ONE SharedCore over the dataset file and runs each param block
    // on its own thread + reader Session (the slice-7 convergence, session.md §2.4/§10) —
    // every Session shares the core's committed snapshot + buffer pool and reads without
    // blocking (§3). The first pass warms the shared pool; the second is wall-clock-timed.
    fn concurrent_read(
        &mut self,
        sql: &str,
        warm: Vec<Vec<Vec<Arg>>>,
        meas: Vec<Vec<Vec<Arg>>>,
        expect_rows: Option<usize>,
    ) -> BoxResult<Option<ConcurrentOutcome>> {
        let path = format!("{}/{}.jed", self.data_dir, self.dataset);
        let core = SharedCore::open(&path).map_err(|e| e.to_string())?;
        let sql = sql.to_string();

        // Pass 1 — warmup, untimed.
        let warm_handles: Vec<_> = warm
            .into_iter()
            .map(|block| {
                let core = core.clone();
                let sql = sql.clone();
                std::thread::spawn(move || -> Result<(), String> {
                    let mut sess = core.read_session();
                    for args in &block {
                        reader_query(&mut sess, &sql, args, None)?;
                    }
                    sess.close();
                    Ok(())
                })
            })
            .collect();
        for h in warm_handles {
            h.join()
                .map_err(|_| "warmup reader panicked".to_string())??;
        }

        // Pass 2 — measured, timed by wall clock around the spawn/join of all readers.
        let start = Instant::now();
        let handles: Vec<_> = meas
            .into_iter()
            .map(|block| {
                let core = core.clone();
                let sql = sql.clone();
                std::thread::spawn(move || -> Result<(String, Vec<i64>, i64), String> {
                    let mut sess = core.read_session();
                    let mut sum = Checksum::new();
                    let mut elapsed = Vec::with_capacity(block.len());
                    let mut rows = 0i64;
                    for args in &block {
                        let t0 = Instant::now();
                        let n = reader_query(&mut sess, &sql, args, Some(&mut sum))?;
                        elapsed.push(t0.elapsed().as_nanos() as i64);
                        rows += n as i64;
                        if let Some(exp) = expect_rows
                            && n != exp
                        {
                            return Err(format!("expected {exp} rows per iteration, got {n}"));
                        }
                    }
                    sess.close();
                    Ok((sum.hex(), elapsed, rows))
                })
            })
            .collect();

        // Join in spawn (= reader-index) order so block_hexes folds in that order (§6).
        let mut block_hexes = Vec::new();
        let mut elapsed = Vec::new();
        let mut rows_total = 0i64;
        for h in handles {
            let (hex, el, rows) = h.join().map_err(|_| "reader panicked".to_string())??;
            block_hexes.push(hex);
            elapsed.extend(el);
            rows_total += rows;
        }
        let wall_ns = start.elapsed().as_nanos() as i64;
        Ok(Some(ConcurrentOutcome {
            block_hexes,
            elapsed,
            rows_total,
            wall_ns,
        }))
    }
}

// reader_query runs one query through a reader Session, re-parsing the SQL each call (the
// host session API has no prepared-statement form; benchmarks.md §8.1), folding rows into
// sum when measuring.
fn reader_query(
    sess: &mut Session,
    sql: &str,
    args: &[Arg],
    sum: Option<&mut Checksum>,
) -> Result<usize, String> {
    let params = bind_args(args);
    let rows = sess.query(sql, &params).map_err(|e| e.to_string())?;
    let mut n = 0;
    match sum {
        None => {
            for _ in rows {
                n += 1;
            }
        }
        Some(sum) => {
            for row in rows {
                n += 1;
                for v in row {
                    match v {
                        Value::Null => sum.null(),
                        Value::Int(x) => sum.int(x),
                        Value::Text(s) => sum.text(&s),
                        other => return Err(format!("unexpected result value {other:?}")),
                    }
                }
                sum.end_row();
            }
        }
    }
    Ok(n)
}
